discovery_engine/lib/src/infrastructure/assets/native/data_provider.dart

```diff
@@ -65,36 +67,55 @@ class NativeDataProvider extends DataProvider {
   Future<String> _getData(Asset asset) async {
     final filePath =
         DataProvider.joinPaths([storageDirectoryPath, asset.urlSuffix]);
-    final assetFile = File(filePath);
-    final diskDirPath = assetFile.parent.path;
-    await Directory(diskDirPath).create(recursive: true);
-
-    // Only write the data on disk if the file does not exist or the checksum does not match.
-    // The last check is useful in case the app is closed before we can finish to write,
-    // and it can be also useful during development to test with different models.
-    var doesExist = assetFile.existsSync();
-
-    if (doesExist &&
-        !await _verifyChecksum(assetFile, asset.checksum.checksumAsHex)) {
-      await assetFile.delete();
-      doesExist = false;
-    }
+    final assetFileRef = File(filePath);
+    await Directory(assetFileRef.parent.path).create(recursive: true);
 
-    if (!doesExist) {
-      final bytes = await assetFetcher.fetchAsset(
-        asset,
-        onFetched: assetReporter.assetFetched,
-      );
-      await assetFile.writeAsBytes(bytes, flush: true);
-    } else if (asset.fragments.isEmpty) {
-      assetReporter.assetFetched(asset.urlSuffix);
-    } else {
+    // if the file exists it means it was verified and has a proper checksum
+    if (assetFileRef.existsSync()) {
+      if (asset.fragments.isEmpty) {
+        assetReporter.assetFetched(asset.urlSuffix);
+      }
       for (final fragment in asset.fragments) {
         assetReporter.assetFetched(fragment.urlSuffix);
       }
+      return assetFileRef.path;
+    }
+
+    // if the file doesn't exist we try to look for the temp file copied from
+    // bundled assets, waiting for checksum verification
+    final tmpFileRef = File('$filePath.$tmpFileExt');
+    final doesTmpFileExist = tmpFileRef.existsSync();
```

Comment:
There a race condition between this check and when we open the file with `_verifyChecksum`. If the file is not there an exception will be raised and we can use to to clean the temp file.